From 881eadabe71fa78c081eda3cd5701768f3778a21 Mon Sep 17 00:00:00 2001
From: Johan Hovold <johan@kernel.org>
Date: Mon, 28 Nov 2016 19:25:09 +0100
Subject: net: dsa: slave: fix fixed-link phydev leaks

From: Johan Hovold <johan@kernel.org>

commit 881eadabe71fa78c081eda3cd5701768f3778a21 upstream.

Make sure to deregister and free any fixed-link PHY registered using
of_phy_register_fixed_link() on slave-setup errors and on slave destroy.

Fixes: 0d8bcdd383b8 ("net: dsa: allow for more complex PHY setups")
Signed-off-by: Johan Hovold <johan@kernel.org>
Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

---
 net/dsa/slave.c |    2 ++
 1 file changed, 2 insertions(+)

--- a/net/dsa/slave.c
+++ b/net/dsa/slave.c
@@ -1083,6 +1083,8 @@ static int dsa_slave_phy_setup(struct ds
 		ret = dsa_slave_phy_connect(p, slave_dev, p->port);
 		if (ret) {
 			netdev_err(slave_dev, "failed to connect to port %d: %d\n", p->port, ret);
+			if (phy_is_fixed)
+				of_phy_deregister_fixed_link(port_dn);
 			return ret;
 		}
 	} else {
